File: python/src/ensoul/resources/health.py
"""Health resource for the Ensoul SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ensoul.http import AsyncHTTPClient, SyncHTTPClient

__all__ = [
    "Health",
    "AsyncHealth",
]


class Health:
    """Synchronous health resource.

    Health endpoints live at /health (no /v1/ prefix).
    Uses get_raw() to skip /v1/ normalization while retaining auth and SDK error mapping.
    """

    def __init__(self, client: SyncHTTPClient) -> None:
        self._client = client

    def check(self) -> dict:
        """GET /health — general health check."""
        return self._client.get_raw("/health").json()

    def ready(self) -> dict:
        """GET /health/ready — readiness check."""
        return self._client.get_raw("/health/ready").json()

    def live(self) -> dict:
        """GET /health/live — liveness check."""
        return self._client.get_raw("/health/live").json()


class AsyncHealth:
    """Asynchronous health resource.

    Health endpoints live at /health (no /v1/ prefix).
    Uses get_raw() to skip /v1/ normalization while retaining auth and SDK error mapping.
    """

    def __init__(self, client: AsyncHTTPClient) -> None:
        self._client = client

    async def check(self) -> dict:
        """GET /health — general health check."""
        return (await self._client.get_raw("/health")).json()

    async def ready(self) -> dict:
        """GET /health/ready — readiness check."""
        return (await self._client.get_raw("/health/ready")).json()

    async def live(self) -> dict:
        """GET /health/live — liveness check."""
        return (await self._client.get_raw("/health/live")).json()

File: swift/Sources/Ensoul/Resources/Health.swift
/// Health resource for the Ensoul Swift SDK.
///
/// Wraps the `/health` family of endpoints. These live **outside** the
/// versioned `/v1/` API root and are used for infrastructure health probes
/// (load balancer readiness, Kubernetes liveness, etc.).
///
/// `getRaw` is used instead of `get` to skip the `/v1/` path-normalisation
/// step that the HTTPClient applies by default.
///
/// Example:
/// ```swift
/// let status = try await client.health.check()
/// print(status["status"] ?? "unknown")
///
/// let ready = try await client.health.ready()
/// print(ready["ready"] ?? false)
/// ```
import Foundation

// MARK: - Health

@available(iOS 15.0, macOS 12.0, *)
public class Health {
    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    // MARK: - Check

    /// GET /health
    ///
    /// General health check. Returns the service status and any component
    /// health information as a raw JSON dictionary.
    public func check() async throws -> [String: Any] {
        let (data, _) = try await client.getRaw("/health")
        return try Self.jsonObject(from: data)
    }

    // MARK: - Ready

    /// GET /health/ready
    ///
    /// Readiness probe — returns `200` when the service is ready to accept
    /// traffic (all dependencies are up). Returns a non-2xx status (and throws)
    /// when the service is not yet ready.
    public func ready() async throws -> [String: Any] {
        let (data, _) = try await client.getRaw("/health/ready")
        return try Self.jsonObject(from: data)
    }

    // MARK: - Live

    /// GET /health/live
    ///
    /// Liveness probe — returns `200` when the process is alive. Use this as
    /// the Kubernetes liveness check; a non-2xx response indicates the pod
    /// should be restarted.
    public func live() async throws -> [String: Any] {
        let (data, _) = try await client.getRaw("/health/live")
        return try Self.jsonObject(from: data)
    }

    // MARK: - Private helpers

    private static func jsonObject(from data: Data) throws -> [String: Any] {
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw EnsoulAPIError(
                statusCode: 200,
                error: "ParseError",
                message: "Expected a JSON object in health response"
            )
        }
        return dict
    }
}
